client/db/src/storage_cache/key_next.rs

```diff
@@ -0,0 +1,865 @@
+// This file is part of Substrate.
+
+// Copyright (C) 2019-2021 Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: GPL-3.0-or-later WITH Classpath-exception-2.0
+
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+//! Structure and implementation used for caching `next_key` calls.
+//! This uses ordered mapping with key intervals, see `CachedInterval`.
+
+
+use std::collections::{HashMap, BTreeMap};
+use sp_core::storage::ChildInfo;
+use super::{EstimateSize,
+};
+
+pub(super) struct LRUOrderedKeys<K> {
+	/// We use a BTreeMap for storage internally.
+	intervals: BTreeMap<K, Box<KeyOrderedEntry<K>>>,
+	/// Intervals for child storages.
+	child_intervals: HashMap<Vec<u8>, BTreeMap<K, Box<KeyOrderedEntry<K>>>>,
+	/// Current total size of contents.
+	used_size: usize,
+	/// Limit size of contents.
+	limit: usize,
+	/// Dummy `KeyOrderedEntry` containing `next` pointer
+	/// as the oldest entry.
+	/// `prev` pointer is used as the lru entry, meaning
+	/// if `prev` equals to `next` the lru structure is empty.
+	lru_bound: Box<KeyOrderedEntry<K>>,
+}
+
+#[derive(Default)]
+pub(super) struct LocalOrderedKeys<K: Ord> {
+	/// We use a BTreeMap for storage internally.
+	intervals: BTreeMap<K, CachedInterval>,
+	/// Intervals for child storages.
+	child_intervals: HashMap<Vec<u8>, BTreeMap<K, CachedInterval>>,
+}
+	
+struct KeyOrderedEntry<K> {
+	/// Entry accessed before.
+	prev: *mut KeyOrderedEntry<K>,
+	/// Entry access after.
+	next: *mut KeyOrderedEntry<K>,
+	/// Used to remove from btreemap.
+	/// Specialized lru struct would not need it.
+	key: K,
+	/// When intervals are in child cache (also only use
+	/// to remove from cache).
+	child_storage_key: Option<Vec<u8>>,
+	/// Actual content.
+	state: CachedInterval,
```

Comment:
Can this be calculated from `self.prev.is_null()` and `self.next.is_null()`?

Reply:
`prev` and `next` are lru pointers (can't be null), when state just indicates the btreemap entry state, can be:
Result of a `next_storage_key` query (CachedInterval::Prev).
Input parameter of a `next_storage_key` (CachedInterval::Next).
Or Both.